README.md
# ornellember

This library exists to convert dates to the ornellember format.

## What is ornellember
ornellember is a joke date format based on the [International Fixed Calendar](https://en.wikipedia.org/wiki/International_Fixed_Calendar#:~:text=7%20External%20links-,Rules,the%20total%20to%20365%20days.). It comprises 13 days that are each 28 days in length, except month 13 (known as M), which contains 29 days on normal years and 30 days on leap years.

Previous proposals for the International Fixed Calendar had assumed that the months would conserve the same names, but this format proposes a new paradigm: each month is referred to by its corresponding letter in the Latin alphabet (A-M). This makes translation easier.

### Holidays
The 29th of M is called N -- which is the next alphabetical letter, but also stands for New. N is the day where all humans can usher in the new year, and breathe in the air of renewal.

Every 4 years, we have the 30th of M, which is called O. Again, O is the next alphabetical letter after N, but more importantly, it stands for Ornellember Day! O is the most special calendar holiday of all. It reminds us of the cyclical nature of life, and of the arbitrary nature of calendars.

### Origins
Ornellember was born from [a late-night tweet](https://twitter.com/zerornella/status/1448764828122759170?s=20). The principle is hardly original, especially considering that the International Fixed Calendar proposal is nearly 150 years old. The name `ornellember` was in response to the question of what to name the 13th month, although the decision to use letters for month names was already taken at that point.

## Getting started
Install the package:
```
npm install ornellember
```

Convert a date:
```
import ornellember from 'ornellember';

const date = new Date('June 25, 1979');
const convertedDate = ornellember(date);

convertedDate.day // 9
convertedDate.month // G
```

## Readable properties
The following properties are available to read:

### `.seconds`
The spare seconds. For example, if it is 9 G, 1979 at 13:45:36, `convertedDate.seconds` would return 36.

### `.minutes`
The spare minutes. For example, if it is 9 G, 1979 at 13:45:36, `convertedDate.minutes` would return 45.

### `.hours`
The spare minutes. For example, if it is 9 G, 1979 at 13:45:36, `convertedDate.hours` would return 13, in 24-hour format.

### `.day`
The day of the month. For example, if it is 9 G, 1979, `convertedDate.day` would return 9.

### `.month`
The month. For example, if it is 9 G, 1979, `convertedDate.month` would return G.

### `.year`
The full year. For example, if it is 9 G, 1979, `convertedDate.year` would return 1979.



## Initializing
ornellember accepts 3 possible inputs to initialize. 
1. a Date object
The ornellember object created will be based on the date passed in.
```
const date = new Date('June 25, 1979');
const convertedDate = ornellember(date);
```

2. nothing
The ornellember object created will be based on the current date.

```
const nowInOrnellember = ornellember();
```

3. an ornellember object
The ornellember object created will be a deep clone of the one passed in.

```
const nowInOrnellember = ornellember();

const copy = ornellember(nowInOrnellember)
```

## Methods
### `.format`
Spits out a string with the format of `day month, year`.

```
const date = new Date('June 25, 1979');
const convertedDate = ornellember(date);

convertedDate.format() \\ '9 G, 1979'
```

### `.add`
We can create a copy of a date, adding units of time from it. Important: the original date will not be altered.

```
const firstDayOfYear = ornellember(new Date('January 1, 2000')); \\ 1 A, 2000
const twoMonthsLater = firstDayOfYear.add(2, 'months'); \\ 1 C, 2000
```

### `.subtract`
We can create a copy of a date, adding units of time from it. Important: the original date will not be altered.

```
const firstDayOfYear = ornellember(new Date('January 1, 2000')); \\ 1 A, 2000
const twoMonthsEarlier = firstDayOfYear.subtract(2, 'months'); \\ 1 L, 1999
```

### `.holiday`
Should return the holiday if the current day is a holiday, and null otherwise. Current supported holidays:
N: 29 M (which is December 31st on non-leap years and December 30 on leap years)
O: 30 M (which is December 31st on leap years)

```
const n = ornellember(new Date('December 31, 1999')).holiday(); \\ 'N'
const o = ornellember(new Date('December 31, 2000')).holiday(); \\ 'O'

```



CHANGELOG.md
11 July 2022 | 0rnella<https://github.com/0rnella>
* fix(readonly): make necessary properties read-only
    commit: c66595ec53c48bc78719591bdc1c3db7cee712e8
    Description: there was a bug where a user could reassign properties in the ornellember object that shouldn't be, for example, could set date.day = something (which wouldn't reflect the raw date). Used Object.assignProperties to make said properties read-only.

* feat(lint): add linting
    commit: dca0e8940cf5ccdd17ef571efd123eecb864a692
    Description: Added Eslint and package.json scripts for it

* cleanup(*): apply linting rules
    commit: eebffac717813c2d87cdce84ebe2fa34b9cc00fe
    Description: Applied eslint rules defined in previous commit and fix errors


08 July 2022 | 0rnella<https://github.com/0rnella>
* another fix to rollup bundle: save output to a named variable called ornellember --  published @2.2.2

08 July 2022 | 0rnella<https://github.com/0rnella>
* Add rollup to bundle output into a file consumable by the browser as-is directly in a script --  published @2.2.0
* fix said bundle by compiling it from the typescript  --  published @2.2.1

lib/bundle.js
var ornellemberBundle=(function(exports){'use strict';var getDaysBeforeCurrentMonth = function (currentMonthIndex, isLeapYear) {
    // then calculate the days of all the months before.
    var daysInMonths = [
        31,
        isLeapYear ? 29 : 28,
        31,
        30,
        31,
        30,
        31,
        31,
        30,
        31,
        30,
        31 // December
    ];
    // how many days before the current month?
    var daysBeforeCurrentMonth = 0;
    for (var i = 0; i < currentMonthIndex; i++) {
        daysBeforeCurrentMonth += daysInMonths[i];
    }
    return daysBeforeCurrentMonth;
};
var isLeapYear = function (date) {
    return (date.getFullYear() % 4) === 0;
};
var getDayNumber = function (date) {
    // first determine if it is a leap year
    var isCurrentLeapYear = isLeapYear(date);
    var currentMonth = date.getMonth();
    // then determine how many days happened before this current month started
    var daysBeforeCurrentMonth = getDaysBeforeCurrentMonth(currentMonth, isCurrentLeapYear);
    var dayOfMonth = date.getDate();
    return daysBeforeCurrentMonth + dayOfMonth;
};
var convertDay = function (dayNumber) {
    var monthLength = 28;
    var monthIndex = Math.floor((dayNumber - 1) / monthLength);
    // make sure there are only 13 months (additional days can be added to month 12)
    if (monthIndex > 12) {
        monthIndex = 12;
    }
    var dayOfMonth = dayNumber - (monthIndex * monthLength);
    var monthLetter = String.fromCharCode(97 + monthIndex).toUpperCase();
    return {
        day: dayOfMonth,
        month: monthLetter
    };
};
/*
    This returns the date a new format that I invented, largely based on the International Fixed Calendar
    https://en.wikipedia.org/wiki/International_Fixed_Calendar
    It's mostly a joke but also I think it is logical and fun
*/
var getOrnellemberDay = function (date) {
    var dayNumber = getDayNumber(date);
    return convertDay(dayNumber);
};/*
    This function adds a certain quantity of a certain Ornellember unit to a Date.
    Two things are important to remember here.
    1. The time units we are adding are Ornellember time units, as they are input by the user.
    2. The original date and the final date are NOT in Ornellember format:
    they are regular Dates from the [Date Constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)
*/
var addToDate = function (originalDate, quantity, unit) {
    var valueToAdd;
    switch (unit) {
        case 'seconds': {
            valueToAdd = quantity * 1000;
            break;
        }
        case 'minutes': {
            valueToAdd = quantity * 1000 * 60;
            break;
        }
        case 'hours': {
            valueToAdd = quantity * 1000 * 60 * 60;
            break;
        }
        case 'days': {
            valueToAdd = quantity * 1000 * 60 * 60 * 24;
            break;
        }
        case 'months': {
            // if we are in the last month of the year 
            if (getDayNumber(originalDate) >= (365 - 29)) {
                if (isLeapYear(originalDate)) {
                    valueToAdd = quantity * 1000 * 60 * 60 * 24 * 30;
                }
                else {
                    valueToAdd = quantity * 1000 * 60 * 60 * 24 * 29;
                }
            }
            else {
                valueToAdd = quantity * 1000 * 60 * 60 * 24 * 28;
            }
            break;
        }
        default: {
            valueToAdd = 0;
            break;
        }
    }
    var newDateValue = originalDate.valueOf() + valueToAdd;
    var newDate = new Date(newDateValue);
    if (unit === 'years') {
        newDate.setFullYear(originalDate.getFullYear() + quantity);
    }
    return newDate;
};var Ornellember = /** @class */ (function () {
    function Ornellember(date) {
        var rawDate = date;
        var converted = getOrnellemberDay(rawDate);
        Object.defineProperties(this, {
            day: { value: converted.day },
            month: { value: converted.month },
            rawDate: { value: rawDate },
            year: { value: rawDate.getFullYear() },
            hours: { value: rawDate.getHours() },
            minutes: { value: rawDate.getMinutes() },
            milliseconds: { value: rawDate.getMilliseconds() }
        });
    }
    Ornellember.prototype.clone = function () {
        return new Ornellember(this.rawDate);
    };
    Ornellember.prototype.add = function (quantity, unit) {
        var newDate = addToDate(this.rawDate, quantity, unit);
        return new Ornellember(newDate);
    };
    Ornellember.prototype.format = function () {
        return this.day + " " + this.month + ", " + this.year;
    };
    Ornellember.prototype.holiday = function () {
        if (this.day === 29)
            return 'N';
        if (this.day === 30)
            return 'O';
        return null;
    };
    Ornellember.prototype.subtract = function (quantity, unit) {
        return this.add(-quantity, unit);
    };
    return Ornellember;
}());
var index = (function (date) {
    if (date instanceof Date) {
        return new Ornellember(date);
    }
    if (date instanceof Ornellember) {
        return date.clone();
    }
    if (!date) {
        return new Ornellember(new Date());
    }
    throw new Error('Unexpected input. Accepted types are a Date, an Ornellember instance, or undefined');
});exports.Ornellember=Ornellember;exports["default"]=index;Object.defineProperty(exports,'__esModule',{value:true});return exports;})({});